builtin/client/autocompletion.lua

```diff
@@ -0,0 +1,83 @@
+local function simple_list_iter(t)
+	local n = 0
+	return function()
+		n = n+1
+		return t[n]
+	end
+end
+
+-- useful to short chat autocompletion code
+function core.autocomplete_word(start, t, iter)
+	if not iter then
+		if t[1] then
+			iter = simple_list_iter
+		else
+			iter = pairs
+		end
+	end
+
+	local start_len = #start
+	local possible_words,n = {},0
+	for word in iter(t) do
+		if word:sub(1, start_len) == start then
+			n = n+1
+			possible_words[n] = word
+		end
+	end
+
+	if n < 2 then
+		return possible_words[1]
+	end
+
+	table.sort(possible_words)
+	local first_word = possible_words[1]
+	local last_word = possible_words[#possible_words]
+	local l = start_len
+	for i = l+1, math.min(#first_word, #last_word) do
+		if first_word:sub(i, i) ~= last_word:sub(i, i) then
+			break
+		end
+		l = l+1
+	end
+
+	if l ~= start_len then
+		return first_word:sub(1, l), possible_words
+	end
+
+	return possible_words
```

Comment:
The function may return either (string, table) or (table). This is bad Lua practice because you have to check the return values for type and have to reorder values, instead of just checking 'if replacement then'
IMO it should return (nil, table) in case there are multiple possibilities that have nothing in common. This makes handling code a lot easier.
Please change line to
return nil, possible_words